### PR TITLE
Implement wp.data.__experimentalResolveSelect

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
We already had two private implementations of a "resolveSelect" and the entities on the widget screen also need this.
This PR proposes a generic wp.data.__experimentalResolveSelect util that returns a promise resolving when the selector resolution is finished.

## How has this been tested?
I created a new post.
I oppened the browser console and pasted: `wp.data.__experimentalResolveSelect('core').getPostType( 'page' ).then(console.log);`.
I verified after a small delay the page post type was printed in the console.

### CODE REVIEW

**Comment by nerrad** [comment]
I think this might be better living in the `@wordpress/data-controls` package?

File: packages/data/package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
 		"equivalent-key-map": "^0.2.2",
 		"is-promise": "^2.1.0",
 		"lodash": "^4.17.14",
+		"memize": "^1.0.5",
```
Comment (epiqueras) [inline packages/data/package.json]: Did you use `lerna` to add this?

I ask because the root dep and lockfiles are not updated.

File: packages/data/src/index.js [hunk 2]
```diff
@@ -91,6 +91,25 @@ export { combineReducers };
  */
 export const select = defaultRegistry.select;
 
+/**
+ * Given the name of a registered store, returns an object that matches the store selectors.
```
Comment (epiqueras) [inline packages/data/src/index.js]: This can be reworded:

"Given the name of a registered store, returns an object containing the store's selectors pre-bound to state so that you only need to supply additional arguments, and modified so that they return promises that resolve to their eventual values, after any resolvers have ran."

File: packages/data/src/index.js [hunk 3]
```diff
@@ -91,6 +91,25 @@ export { combineReducers };
...
+ * The functions return a promise that resolves when the selector resolution is finished.
+ * The selector functions are been pre-bound to pass the current state automatically.
+ * As a consumer, you need only pass arguments of the selector, if applicable.
+ *
+ * @param {string} name Store name
```
Comment (epiqueras) [inline packages/data/src/index.js]: Missing period at the end.

File: packages/data/src/index.js [hunk 4]
```diff
@@ -91,6 +91,25 @@ export { combineReducers };
...
+ *
+ * @example
+ * ```js
+ * const { __experimentalResolveSelect } = wp.data;
+ *
+ * select( 'my-shop' ).getPrice( 'hammer' ).then(console.log)
```
Comment (epiqueras) [inline packages/data/src/index.js]:
```suggestion
 * __experimentalResolveSelect( 'my-shop' ).getPrice( 'hammer' ).then(console.log)
```

File: packages/data/src/index.js [hunk 5]
```diff
@@ -91,6 +91,25 @@ export { combineReducers };
...
+ * ```
+ *
+ * @return {Object} Object containing the store's selectors.
```
Comment (epiqueras) [inline packages/data/src/index.js]:
```suggestion
 * @return {Object} Object containing the store's promise-wrapped selectors.
```

File: packages/data/src/registry.js [hunk 6]
```diff
@@ -88,6 +90,59 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
...
+					};
+				}
+			);
+		},
+		{ maxSize: 1 }
+	);
+
+	/**
+	 * Calls a selector given the current state and extra arguments and
+	 * return a promise that resolves when the selector resolution is finished.
```
Comment (epiqueras) [inline packages/data/src/registry.js]:
```suggestion
	 * returns a promise that resolves when the selector resolution is finished.
```

File: packages/data/src/registry.js [hunk 7]
```diff
@@ -88,6 +90,59 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
...
+	 *
+	 * @param {string} reducerKey Part of the state shape to register the
+	 *                            selectors for.
+	 *
+	 * @return {Promise} Promise resolving to the selector returned value.
```
Comment (epiqueras) [inline packages/data/src/registry.js]: This returns an object not a promise.

File: packages/data/src/registry.js [hunk 8]
```diff
@@ -143,6 +198,7 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	let registry = {
 		registerGenericStore,
 		stores,
+		__experimentalResolveSelect,
```
Comment (epiqueras) [inline packages/data/src/registry.js]: Can we move this under `select`?

File: packages/editor/src/store/effects/utils.js [hunk 9]
```diff
@@ -13,21 +13,5 @@ import { select, subscribe } from '@wordpress/data';
  * @return {Promise} Selector result.
  */
 export function resolveSelector( namespace, selectorName, ...args ) {
-	return new Promise( ( resolve ) => {
-		const hasFinished = () => select( 'core/data' ).hasFinishedResolution( namespace, selectorName, args );
-		const getResult = () => select( namespace )[ selectorName ].apply( null, args );
```
Comment (epiqueras) [inline packages/editor/src/store/effects/utils.js]: I think we should just replace the usage directly in this case, and in the widgets package too.

**Review by epiqueras — DISMISSED** [review]

**Comment by nerrad** [comment]
And actually, how does this differ from the `select` control found in `@wordpress/data-controls` which will automatically do either `resolveSelect` or `select` as needed?

https://github.com/WordPress/gutenberg/blob/aedd241bae2430099c66508349beeecc9ce218b2/packages/data-controls/src/index.js#L160-L166

**Comment by jorgefilipecosta** [comment]
> And actually, how does this differ from the select control found in @wordpress/data-controls which will automatically do either resolveSelect or select as needed?

Hi @nerrad, we can use the control in our actions, but how can I apply a resolve select outside of a store context (e.g., in the browser console)?
Or how can my external javascript code request the page post type and do an action after it is available?
E.g: how can we do with controls something like?
```
wp.data.__experimentalResolveSelect('core').getPostType( 'page' ).then(console.log);
```


I think data-controls don't allow this type of use case. If I am missing something, please let me know. If we advance with this PR, we can use the experimental function in the select control and avoid duplicate code.

**Comment by nerrad** [comment]
> but how can I apply a resolve select outside of a store context (e.g., in the browser console)?

Ahh, now I understand the difference here.  Sorry about not picking that up right away.  I saw the implementation in an `effects.js` file and assumed it was only intended for use as a side-effect on an action (we probably should be converting places we are using effects over to use actions/action generators at some point).

The question I have at this point is whether we _should_ expose selectors this way?  So a bit of a thought exercise going through my mind here: Certainly, there's benefits for quickly debugging in the browser console, but do we want to potentially encourage use of `resolveSelect` outside of a context that reacts to changes in state when resolution is complete?  Or when it's not needed.  For example, are there any concerns about people potentially doing something like this:

```jsx
import { resolveSelect } from '@wordpress/data';
export default withSelect( ( select, props ) => {
    let type;
    resolveSelect( 'core' ).getPostType( 'page' ).then( ( postType ) => {
        type = postType;
    } );
    return {  type };
} )( MyComponent );
```

I realize that's a bit contrived, but I wouldn't be surprised to see it in the wild, or something like:

```jsx
import { resolveSelect } from '@wordpress/data';

export PostTitle = ( { id } ) => {
    let title = null;
    resolveSelect( 'core' ).getPost( id ).then( ( post ) => {
        title = <span>post.title.raw</span>
    } );
   return title;
};
```

I realize bad patterns can end up getting used with other exposed methods as well so this may be overthinking it.  I also realize that this is prefixed by `__experimental` so it's pretty clear this isn't intended as an official public api yet.

The above is not intended to be a blocker to this pull - just some thoughts that came to mind as I'm thinking through some of the ramifications of this.

File: packages/data/package.json (hunk 1 above)
Comment (jorgefilipecosta) [reply packages/data/package.json]: No, I did not use Lerna, but I already updated the lock file.

File: packages/data/src/registry.js (hunk 8 above)
Comment (jorgefilipecosta) [reply packages/data/src/registry.js]: Hi @epiqueras,
You mean to have something like?
wp.data.select.__experimatalResolve( storeName )

**Comment by jorgefilipecosta** [comment]
Hi @nerrad thank you for sharing your thoughts.

> The question I have at this point is whether we should expose selectors this way? So a bit of a thought exercise going through my mind here: Certainly, there's benefits for quickly debugging in the browser console, but do we want to potentially encourage use of resolveSelect outside of a context that reacts to changes in state when resolution is complete?

There are some cases outside a React context where we need to issue a request and then do something after the request finishes.
For example, in PR https://github.com/WordPress/gutenberg/pull/17319 we needed that for integration with customizer (waitForSelectValue function).
If we don't provide something like this, plugin developers that have a similar need will probably end up using wp.apiFetch.

Most non-reactive scenarios can already be done with other utils/functions we provide.
That said I am also not totally sure about what will happen that's why I think we should keep the util experimental.

File: packages/data/src/registry.js (hunk 8 above)
Comment (epiqueras) [reply packages/data/src/registry.js]: I meant to move the line under `select,` to match the order of the rest of the code.

**Comment by jorgefilipecosta** [comment]
Hi @epiqueras, your review was addressed. It is ready for a new review :)

File: packages/data-controls/src/test/index.js [hunk 10]
```diff
@@ -31,9 +31,21 @@ describe( 'controls', () => {
...
 		selectorWithFalseResolver.hasResolver = false;
 		const hasFinishedResolution = jest.fn();
-		const unsubscribe = jest.fn();
-		let subscribedCallback;
+		const resolveSelectSelectorWithResolver = jest.fn();
+		const resolveSelectselectorWithUndefinedResolver = jest.fn();
+		const resolveSelectselectorWithFalseResolver = jest.fn();
+
 		const registryMock = {
+			__experimentalResolveSelect: ( storeKey ) => {
```
Comment (epiqueras) [inline packages/data-controls/src/test/index.js]: This makes the test hard to follow.

`__experimentalResolveSelect` and `select` should share the same `mockStore` and as many mock functions as possible.

File: packages/data-controls/src/test/index.js [hunk 11]
```diff
@@ -62,13 +68,13 @@ describe( 'controls', () => {
...
 			selectorWithResolver.mockReturnValue( 'resolved' );
 			hasFinishedResolution.mockReturnValue( false );
+			resolveSelectSelectorWithResolver.mockResolvedValue( 'resolved' );
 		} );
 		afterEach( () => {
 			selectorWithUndefinedResolver.mockClear();
 			selectorWithResolver.mockClear();
 			selectorWithFalseResolver.mockClear();
 			hasFinishedResolution.mockClear();
-			unsubscribe.mockClear();
```
Comment (epiqueras) [inline packages/data-controls/src/test/index.js]: Shouldn't we clear the new mocks here as well?

**Review by epiqueras — CHANGES_REQUESTED** [review]

**Comment by epiqueras** [comment]
I think you missed https://github.com/WordPress/gutenberg/pull/17558#discussion_r329712289.

File: packages/data-controls/src/test/index.js (hunk 11 above)
Comment (jorgefilipecosta) [reply packages/data-controls/src/test/index.js]: We are using the same mocks now so it is not needed anymore.

File: packages/data-controls/src/test/index.js (hunk 10 above)
Comment (jorgefilipecosta) [reply packages/data-controls/src/test/index.js]: Uodated 👍

**Comment by jorgefilipecosta** [comment]
Hi @epiqueras, thank you for the rereview. I applied you last suggestions and fixed the missing change https://github.com/WordPress/gutenberg/pull/17558#discussion_r329712289 by moving resolveSelect line under the select line.

File: packages/data-controls/src/test/index.js [hunk 12]
```diff
@@ -62,13 +63,13 @@ describe( 'controls', () => {
 			selectorWithFalseResolver.mockReturnValue( 'bar' );
 			selectorWithResolver.mockReturnValue( 'resolved' );
```
Comment (epiqueras) [inline packages/data-controls/src/test/index.js]: I think we can remove this line now.

**Review by epiqueras — APPROVED** [review]

File: packages/data-controls/src/test/index.js [hunk 13]
```diff
@@ -62,13 +63,13 @@ describe( 'controls', () => {
...
 			hasFinishedResolution.mockReturnValue( false );
+			selectorWithResolver.mockResolvedValue( 'resolved' );
```
Comment (epiqueras) [inline packages/data-controls/src/test/index.js]:
```suggestion
```

File: packages/data-controls/src/test/index.js (hunk 12 above)
Comment (epiqueras) [inline packages/data-controls/src/test/index.js]:
```suggestion
			selectorWithResolver.mockResolvedValue( 'resolved' );
```

File: packages/data/src/registry.js [hunk 14]
```diff
@@ -88,6 +90,61 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
...
+						} );
+					};
+				}
+			);
+		},
+		{ maxSize: 1 }
+	);
+
+	/**
+ 	 * Given the name of a registered store, returns an object containing the store's
```
Comment (aduth) [inline packages/data/src/registry.js]: There is an ESLint warning here caused by mixed whitespace at the beginning of the line.

```
/Users/andrew/Documents/Code/gutenberg/packages/data/src/registry.js
  134:0  warning  Expected JSDoc block to be aligned  jsdoc/check-alignment
```

See also: #18458
